### PR TITLE
Enable model training and NNCF in mmdet

### PR DESCRIPTION
**Pull request by cih9088**
### Previous PR: https://github.com/openvinotoolkit/training_extensions/pull/1316

- Most of the monkey patches are not to add functional features but just to block the trace of its methods/functions in NNCF.
- Add trainable and NNCF optimisable YOLOX
- Add trainable and NNCF optimisable SSD
- Rearange nncf adapter
- Add instance-segmentation test code
- Add trainable and NNCF optimisable Mask R-CNN

### CODE REVIEW

File: otx/algorithms/detection/adapters/mmdet/utils/config_utils.py [hunk 1]
```diff
@@ -193,7 +193,6 @@ def prepare_for_training(
     config.data.val.otx_dataset = val_dataset
     patch_adaptive_repeat_dataset(config, len(train_dataset))
     config.custom_hooks.append({"type": "OTXProgressHook", "time_monitor": time_monitor, "verbose": True})
-    config.log_config.hooks.append({"type": "OTXLoggerHook", "curves": learning_curves})
```
Comment (cih9088) [inline otx/algorithms/detection/adapters/mmdet/utils/config_utils.py:196]: https://github.com/openvinotoolkit/training_extensions/pull/1316#discussion_r1013812675

File: otx/algorithms/detection/adapters/mmdet/utils/config_utils.py [hunk 2]
```diff
@@ -204,7 +203,7 @@ def set_data_classes(config: Config, labels: List[LabelEntity]):
     for subset in ("train", "val", "test"):
         cfg = get_data_cfg(config, subset)
         cfg.labels = labels
-        config.data[subset].labels = labels
+        #  config.data[subset].labels = labels
 
     # Set proper number of classes in model's detection heads.
```
Comment (cih9088) [inline otx/algorithms/detection/adapters/mmdet/utils/config_utils.py:208]: https://github.com/openvinotoolkit/training_extensions/pull/1316#discussion_r1011487701

File: otx/algorithms/detection/adapters/mmdet/nncf/compression.py [hunk 3]
```diff
@@ -236,6 +205,8 @@ def model_eval_fn(model):
         nncf_compress_postprocessing = cfg.get('nncf_compress_postprocessing')
         logger.debug('set should_compress_postprocessing='f'{nncf_compress_postprocessing}')
     else:
+        # TODO: Do we have to keep this configuration?
+        # This configuration is not enabled in forked mmdetection library in the first place
```
Comment (cih9088) [inline otx/algorithms/detection/adapters/mmdet/nncf/compression.py:209]: https://github.com/openvinotoolkit/training_extensions/pull/1316#discussion_r1013805726

File: otx/algorithms/detection/configs/detection/mobilenetv2_ssd/model.py [hunk 4]
```diff
@@ -69,9 +74,6 @@
             target_means=(0.0, 0.0, 0.0, 0.0),
             target_stds=(0.1, 0.1, 0.2, 0.2),
         ),
-        depthwise_heads=True,
-        depthwise_heads_activations="relu",
-        loss_balancing=False,
     ),
     train_cfg=dict(
         assigner=dict(
```
Comment (cih9088) [inline otx/algorithms/detection/configs/detection/mobilenetv2_ssd/model.py:79]: https://github.com/openvinotoolkit/training_extensions/pull/1316#discussion_r1013811501

File: otx/algorithms/detection/tasks/nncf.py [hunk 5]
```diff
@@ -311,8 +315,7 @@ def optimize(
             loading_stage_progress_percentage=5,
             initialization_stage_progress_percentage=5,
         )
-        learning_curves = DefaultDict(OTXLoggerHook.Curve)  # type: DefaultDict
```
Comment (cih9088) [inline otx/algorithms/detection/tasks/nncf.py:314]: https://github.com/openvinotoolkit/training_extensions/pull/1316#discussion_r1013812675

File: otx/algorithms/detection/tasks/train.py [hunk 6]
```diff
@@ -255,3 +245,18 @@ def _generate_training_metrics(learning_curves, scores) -> Iterable[MetricsGroup
         )
 
         return output
+
+    def _initialize_post_hook(self):
```
Comment (cih9088) [inline otx/algorithms/detection/tasks/train.py:249]: https://github.com/openvinotoolkit/training_extensions/pull/1316#discussion_r1013814074

**Comment by cih9088** [comment]
https://github.com/openvinotoolkit/training_extensions/pull/1316#issuecomment-1305004979

File: otx/algorithms/detection/tasks/nncf.py (hunk 5 above)
Comment (cih9088) [reply otx/algorithms/detection/tasks/nncf.py:314]: Duplicated

**Comment by cih9088** [comment]
Additional commits:
- [fix: mismatches](https://github.com/openvinotoolkit/training_extensions/pull/1355/commits/f0f934f1d6c55762941cae5e9e5f4572528d997d)
- [fix: update pipeline for wrapper](https://github.com/openvinotoolkit/training_extensions/pull/1355/commits/5b6515f47056fada250783d8c086ef6aab8f6f56)

**Comment by goodsong81** [comment]
run ote-test
